### ISSUE TITLE
Backed access public API

### ISSUE THREAD

**Issue by ivirshup** [theislab/anndata#690]
We would like to expose an API for working with completley backed `AnnData` (and `MuData`) objects.

# Current state

Currently we only expose limited functionality for this, `backed` mode, `AnnLoader` and `read_elem`/ `write_elem`.

Backed mode support partial access to backed `X` objects. `AnnLoader` allows per element control and partial access for a subset of the API. `read_elem` and `write_elem` provide low level interfaces to read single elements of an `AnnData` object.

# What we'd like to have

The ability to read a subset of any element of an AnnData object. That is, given an on disk anndata only read statistics from `obs` for cells of a particular cell type. This could later be expanded to a higher level backed API using `dask` for a fully lazy `AnnData` object.

This has been partially implemented internally with `read_elem_partial`, but that has not been given a consistent API or use. Solving this issue could consist of just coming up with, and fully implementing that API.

(Ping @Koncopd @davidsebfischer)

**Comment by Koncopd** [comment]
Agree, partial access would be really useful. Also the current backed mode isn't just incomplete, it is partially broken wrt subsets etc.

**Comment by kaizhang** [comment]
Does this allow fully backed anndata? For example, do not load `.obsm`, `.varm`, `.uns` into memory. I'm using `.obsm` to store large spare matrices. So a lazy `.obsm` will be very useful to me.
